File: src/main/java/com/evergreenclient/hudmod/elements/impl/ElementImage.java
/*
 * Copyright (C) Evergreen [2020 - 2021]
 * This program comes with ABSOLUTELY NO WARRANTY
 * This is free software, and you are welcome to redistribute it
 * under the certain conditions that can be found here
 * https://www.gnu.org/licenses/lgpl-3.0.en.html
 *
 * If you have any questions or concerns, please create
 * an issue on the github page that can be found here
 * https://github.com/Evergreen-Client/EvergreenHUD
 *
 * If you have a private concern, please contact
 * isXander @ business.isxander@gmail.com
 */

package com.evergreenclient.hudmod.elements.impl;

import club.sk1er.mods.core.gui.notification.Notifications;
import com.evergreenclient.hudmod.EvergreenHUD;
import com.evergreenclient.hudmod.elements.Element;
import com.evergreenclient.hudmod.gui.elements.BetterGuiSlider;
import com.evergreenclient.hudmod.gui.screens.impl.GuiElementConfig;
import com.evergreenclient.hudmod.settings.impl.*;
import com.evergreenclient.hudmod.utils.*;
import net.apolloclient.utils.GLRenderer;
import net.minecraft.client.gui.ScaledResolution;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.texture.AbstractTexture;
import net.minecraft.client.renderer.texture.ITextureObject;
import net.minecraft.client.renderer.texture.TextureUtil;
import net.minecraft.client.resources.IResourceManager;
import net.minecraft.crash.CrashReport;
import net.minecraft.crash.CrashReportCategory;
import net.minecraft.util.ReportedException;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.client.event.RenderGameOverlayEvent;

import javax.imageio.ImageIO;
import javax.swing.*;
import javax.swing.filechooser.FileNameExtensionFilter;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

public class ElementImage extends Element {

    private static boolean warned = false;
    private static final File imageFile = new File(mc.mcDataDir, "config/evergreenhud/image.png");
    private static final ResourceLocation unknownImage = new ResourceLocation("evergreenhud/textures/unknown.png");
    private static final float imgSize = 60;

    private Dimension imageDimension;
    private final Map<File, ITextureObject> fileTextureMap = new HashMap<>();
    private boolean changed = false;
    private float scaleMod = 1;

    private StringSetting fileLocation;
    public BooleanSetting mirror;
    public ArraySetting rotation;
    public BooleanSetting autoScale;

    @Override
    public GuiElementConfig getElementConfigGui() {
        return new GuiElementConfig(this) {
            @Override
            public void addButtons() {
                super.addButtons();
                this.buttonList = this.buttonList.stream().filter(b -> b.id != 3).collect(Collectors.toList());
                this.buttonList.add(new BetterGuiSlider(3, left(), getRow(0), 120, 20, "Scale: ", "%", 1, 300, element.getPosition().getScale() * 100f, false, true, this, "Controls the size of the element."));
            }
        };
    }

    @Override
    public boolean useAlignmentSetting() {
        return false;
    }

    @Override
    public boolean useTitleSetting() {
        return false;
    }

    @Override
    public boolean useInvertedSetting() {
        return false;
    }

    @Override
    public boolean useBracketsSetting() {
        return false;
    }

    @Override
    public boolean useTextColorSetting() {
        return false;
    }

    @Override
    public boolean useChromaSetting() {
        return false;
    }

    @Override
    public boolean useShadowSetting() {
        return false;
    }

    private File getImageFile() {
        if (fileLocation.get().equals("") || !new File(fileLocation.get()).exists()) {
            if (!imageFile.exists()) {
                copyNullImage();
            }
            fileLocation.set(imageFile.getPath());
        }

        return new File(fileLocation.get());
    }

    private void copyNullImage() {
        try {
            Files.copy(ElementImage.class.getResourceAsStream("/assets/" + unknownImage.getResourceDomain() + "/" + unknownImage.getResourcePath()), imageFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
            BufferedImage img = ImageIO.read(imageFile);
            imageDimension = new Dimension(img.getWidth(), img.getHeight());
        } catch (IOException e) {
            e.printStackTrace();
            throw new ReportedException(CrashReport.makeCrashReport(e, "Failed to copy image."));
        }
    }

    @Override
    public void resetSettings(boolean save) {
        super.resetSettings(save);

        imageFile.delete();
        copyNullImage();
    }

    @Override
    public void initialise() {
        addSettings(fileLocation = new StringSetting("File Path", "", true));

        addSettings(new ButtonSetting("Choose Image", "Picks your desired image to be rendered.", () -> {
            // Add to new thread so the game does not freeze
            new Thread(() -> {
                JFileChooser fileChooser = new JFileChooser();
                fileChooser.setCurrentDirectory(new File(mc.mcDataDir, "config/evergreenhud/"));
                fileChooser.setFileFilter(new FileNameExtensionFilter("Image File", "png", "jpg"));
                int result = fileChooser.showOpenDialog(null);
                if (result == JFileChooser.APPROVE_OPTION) {
                    try {
                        File file = fileChooser.getSelectedFile();
                        BufferedImage image = ImageIO.read(file);
                        fileTextureMap.remove(getImageFile());
                        fileLocation.set(file.getPath());
                        imageDimension = new Dimension(image.getWidth(), image.getHeight());
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                    Notifications.INSTANCE.pushNotification("EvergreenHUD", "You have selected a new image.");
                } else {
                    Notifications.INSTANCE.pushNotification("EvergreenHUD", "You must select an image.");
                }
                changed = true;
            }).start();
        }));
        addSettings(mirror = new BooleanSetting("Mirror", "If the image is flipped horizontally.", false) {
            @Override
            protected boolean onChange(boolean oldValue, boolean newValue) {
                changed = true;
                return true;
            }
        });
        addSettings(rotation = new ArraySetting("Rotation", "How will the image be rotated. (Image should be square for best result.)", 0, "0 deg", "90 deg", "180 deg", "270 deg") {
            @Override
            protected boolean onChange(int currentIndex, int newIndex) {
                changed = true;
                return true;
            }
        });
        addSettings(autoScale = new BooleanSetting("Auto Scale", "Automatically scales your image to a constant size depending on the scale.", true));
    }

    @Override
    protected void onSettingsLoad() {
        try {
            if (getImageFile().equals(imageFile)) {
                if (!EvergreenHUD.getInstance().isFirstLaunch() && !warned) {
                    Notifications.INSTANCE.pushNotification("EvergreenHUD", "Could not find image on one or more of your custom image elements.\nPlease reselect your images.");
                    warned = true;
                }
            }

            BufferedImage img = ImageIO.read(getImageFile());
            imageDimension = new Dimension(img.getWidth(), img.getHeight());
        } catch (IOException e) {
            e.printStackTrace();
            throw new ReportedException(CrashReport.makeCrashReport(e, "An error was encountered."));
        }
    }

    @Override
    protected ElementData metadata() {
        return new ElementData("Custom Image", "Renders an image of your choosing.");
    }

    @Override
    public void render(RenderGameOverlayEvent event) {
        mc.mcProfiler.startSection(getMetadata().getName());
        GlStateManager.pushMatrix();
        float scale = getPosition().getScale();

        if (autoScale.get()) {
            scaleMod = imgSize / Math.min((float) imageDimension.getWidth(), (float) imageDimension.getHeight());
        } else {
            scaleMod = 1;
        }
        Hitbox hitbox = getHitbox(1, scale);
        GLRenderer.drawRectangle(hitbox.x, hitbox.y, hitbox.width, hitbox.height, getBgColor());
        GlStateManager.scale(scale, scale, 1);
        GlStateManager.enableDepth();
        GlStateManager.color(1f, 1f, 1f, 1f);
        bindTexture(getImageFile());
        double width = imageDimension.getWidth();
        double height = imageDimension.getHeight();
        double renderWidth = width * scaleMod;
        double renderHeight = height * scaleMod;

        RenderUtils.drawModalRect(this.getPosition().getRawX(event.resolution) / scale, this.getPosition().getRawY(event.resolution) / scale, 0, 0, imageDimension.getWidth(), imageDimension.getHeight(), renderWidth, renderHeight, imageDimension.getWidth(), imageDimension.getHeight());

        GlStateManager.popMatrix();
        mc.mcProfiler.endSection();
    }

    @Override
    public Hitbox getHitbox(float posScale, float sizeScale) {
        ScaledResolution res = new ScaledResolution(mc);

        float width = (float)imageDimension.getWidth() * sizeScale * scaleMod;
        float height = (float)imageDimension.getHeight() * sizeScale * scaleMod;

        float extraWidth = getPaddingWidth() * sizeScale;
        float extraHeight = getPaddingHeight() * sizeScale;
        float x = getPosition().getRawX(res) / posScale;
        float y = getPosition().getRawY(res) / posScale;

        return new Hitbox(x - extraWidth, y - extraHeight, width + (extraWidth * 2), height + (extraHeight * 2));
    }

    private void bindTexture(File file) {
        ITextureObject texture = fileTextureMap.get(file);
        if (texture == null || changed) {
            texture = new ExternalFileTexture(file, this);
            loadTexture(file, texture);
            changed = false;
        }

        GlStateManager.bindTexture(texture.getGlTextureId());
    }

    private void loadTexture(File file, ITextureObject texture) {
        try {
            texture.loadTexture(mc.getResourceManager());
        } catch (IOException e) {
            logger.warn("Failed to load texture: " + file.getPath(), e);
            texture = TextureUtil.missingTexture;
            this.fileTextureMap.put(file, texture);
        } catch (Throwable throwable) {
            ITextureObject finalTexture = texture;
            CrashReport crash = CrashReport.makeCrashReport(throwable, "Registering texture");
            CrashReportCategory crashreportcategory = crash.makeCategory("Resource location being registered");
            crashreportcategory.addCrashSection("Resource location", file.getPath());
            crashreportcategory.addCrashSectionCallable("Texture object class", () -> finalTexture.getClass().getName());
            throw new ReportedException(crash);
        }
        this.fileTextureMap.put(file, texture);
    }

    @Override
    protected String getValue() {
        return null;
    }

    @Override
    public String getDisplayTitle() {
        return null;
    }

    public static class ExternalFileTexture extends AbstractTexture {

        protected final File textureLocation;
        protected final ElementImage element;

        public ExternalFileTexture(File file, ElementImage element) {
            this.textureLocation = file;
            this.element = element;
        }

        @Override
        public void loadTexture(IResourceManager resourceManager) throws IOException {
            this.deleteGlTexture();
            InputStream is = new FileInputStream(textureLocation);

            BufferedImage img = TextureUtil.readBufferedImage(is);
            if (element.mirror.get())
                img = ImageUtils.flipHorizontally(img);
            img = ImageUtils.rotate(img, element.rotation.getIndex() * 90);

            TextureUtil.uploadTextureImageAllocate(this.getGlTextureId(), img, false, false);
            is.close();
        }
    }

}

File: src/main/kotlin/dev/isxander/evergreenhud/elements/impl/ElementImage.kt
/*
 * EvergreenHUD - A mod to improve your heads-up-display.
 * Copyright (c) isXander [2019 - 2022].
 *
 * This work is licensed under the GPL-3 License.
 * To view a copy of this license, visit https://www.gnu.org/licenses/gpl-3.0.en.html
 */

package dev.isxander.evergreenhud.elements.impl

import dev.isxander.evergreenhud.EvergreenHUD
import dev.isxander.evergreenhud.elements.RenderOrigin
import dev.isxander.evergreenhud.elements.type.BackgroundElement
import dev.isxander.evergreenhud.utils.Color
import dev.isxander.evergreenhud.utils.elementmeta.ElementMeta
import dev.isxander.evergreenhud.utils.mc
import dev.isxander.settxi.impl.boolean
import dev.isxander.settxi.impl.file
import dev.isxander.settxi.impl.float
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.renderer.Tessellator
import net.minecraft.client.renderer.texture.DynamicTexture
import net.minecraft.client.renderer.vertex.DefaultVertexFormats
import net.minecraft.util.ResourceLocation
import java.awt.Dimension
import java.awt.image.BufferedImage
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStream
import javax.imageio.ImageIO


@ElementMeta(id = "evergreenhud:image", name = "Image", category = "Miscallaneous", description = "Display an image file.")
class ElementImage : BackgroundElement() {
    var file by file(EvergreenHUD.dataDir) {
        name = "File"
        category = "Image"
        description = "The file of the image."

        set {
            changed = true
            it
        }
    }

    var autoScale by boolean(true) {
        name = "Auto Scale"
        category = "Image"
        description = "Scale your image so it is the same size, no matter the resolution."
    }

    var rotation by float(0f) {
        range = 0f..360f
        name = "Rotation"
        category = "Image"
        description = "Rotate the image x amount of degrees."
    }

    var mirror by boolean(false) {
        name = "Mirror"
        category = "Image"
        description = "Makes your image look as if you viewed it through a mirror."

        set {
            changed = true
            it
        }
    }

    var changed = true
    private var currentImage: ResourceLocation? = null
    var imageDimension = Dimension(64, 64)
    var scaleMod = 1f
    var imgSize = 64
    val textureName = "evergreen-image-element-${hashCode()}"
    private val unknownImage = ResourceLocation("evergreenhud", "textures/unknown.png")

    override val hitboxWidth: Float
        get() = imageDimension.width * scaleMod
    override val hitboxHeight: Float
        get() = imageDimension.height * scaleMod

    init {
        backgroundColor = Color.black.withAlpha(100)
        paddingLeft = 0f
        paddingRight = 0f
        paddingTop = 0f
        paddingBottom = 0f
    }

    override fun render(renderOrigin: RenderOrigin) {
        if (changed || currentImage == null) {
            // Reload the texture
            if (currentImage != null) mc.textureManager.deleteTexture(currentImage)
            try {
                cacheResourceLocation()
            } catch (e: IOException) {
                e.printStackTrace()
            }
            changed = false
        }

        GlStateManager.pushMatrix()
        val scale = position.scale

        scaleMod = if (autoScale) {
            imgSize / imageDimension.getWidth().toFloat().coerceAtMost(imageDimension.getHeight().toFloat())
        } else {
            1f
        }
        GlStateManager.pushMatrix()
        val hitbox = calculateHitBox(position.scale)
        val bgPivotX: Float = hitbox.x1 + hitbox.width / 2f
        val bgPivotY: Float = hitbox.x1 + hitbox.height / 2f
        GlStateManager.translate(bgPivotX, bgPivotY, 0f)
        GlStateManager.rotate(rotation, 0f, 0f, 1f)
        GlStateManager.translate(-bgPivotX, -bgPivotY, 0f)
        super.render(renderOrigin)
        GlStateManager.popMatrix()
        GlStateManager.scale(scale, scale, 1f)
        GlStateManager.enableDepth()
        GlStateManager.color(1f, 1f, 1f, 1f)
        mc.textureManager.bindTexture(currentImage)
        val width = imageDimension.getWidth()
        val height = imageDimension.getHeight()
        val renderWidth = width * scaleMod
        val renderHeight = height * scaleMod
        val renderX = position.rawX
        val renderY = position.rawY

        val imgPivotX = (renderX + renderWidth.toFloat() / 2f) / scale
        val imgPivotY = (renderY + renderHeight.toFloat() / 2f) / scale
        GlStateManager.translate(imgPivotX, imgPivotY, 0f)
        GlStateManager.rotate(rotation, 0f, 0f, 1f)
        GlStateManager.translate(-imgPivotX, -imgPivotY, 0f)

        drawModalRect(
            renderX / scale,
            renderY / scale,
            0,
            0,
            imageDimension.getWidth(),
            imageDimension.getHeight(),
            renderWidth,
            renderHeight,
            imageDimension.getWidth(),
            imageDimension.getHeight()
        )

        GlStateManager.popMatrix()
    }

    @Throws(IOException::class)
    private fun cacheResourceLocation() {
        val imgFile: File? = getImageFile()
        val `in`: InputStream
        if (imgFile == null) {
            `in` = mc.resourceManager.getResource(unknownImage).inputStream
        } else {
            `in` = FileInputStream(imgFile)
        }
        var img: BufferedImage = ImageIO.read(`in`)
        if (mirror) img = mirror(img)
        currentImage = mc.textureManager.getDynamicTextureLocation(textureName, DynamicTexture(img))
        imageDimension = Dimension(img.width, img.height)
    }

    private fun getImageFile(): File? {
        return if (file.name == "" || file.exists()) {
            null
        } else file
    }

    private fun mirror(image: BufferedImage): BufferedImage {
        val newImage = BufferedImage(image.width, image.height, image.type)
        for (y in 0 until newImage.height) {
            var x1 = 0
            var x2 = newImage.width - 1
            while (x1 < newImage.width) {
                newImage.setRGB(x1, y, image.getRGB(x2, y))
                x1++
                x2--
            }
        }
        return newImage
    }

    fun drawModalRect(
        x: Float,
        y: Float,
        u: Int,
        v: Int,
        uWidth: Double,
        vHeight: Double,
        width: Double,
        height: Double,
        tileWidth: Double,
        tileHeight: Double
    ) {
        val f = 1.0 / tileWidth
        val f1 = 1.0 / tileHeight
        val tessellator = Tessellator.getInstance()
        val worldrenderer = tessellator.worldRenderer
        worldrenderer.begin(7, DefaultVertexFormats.POSITION_TEX)
        worldrenderer.pos(x.toDouble(), y + height, 0.0).tex(u * f, (v + vHeight) * f1).endVertex()
        worldrenderer.pos(x + width, y + height, 0.0).tex((u + uWidth) * f, (v + vHeight) * f1).endVertex()
        worldrenderer.pos(x + width, y.toDouble(), 0.0).tex((u + uWidth) * f, v * f1).endVertex()
        worldrenderer.pos(x.toDouble(), y.toDouble(), 0.0).tex(u * f, v * f1).endVertex()
        tessellator.draw()
    }
}
